### PR TITLE
Test framework update

### PR DESCRIPTION
**Pull request by JonathanBurke**
The test framework is replaced.  The new class to extend is DefaultCheckerTest.  
If you have a custom test, you can always use the test utilities to easily run the framework with a given set of options.

### CODE REVIEW

File: checker/tests/src/tests/IGJTest.java [hunk 1]
```diff
@@ -1,33 +1,45 @@
...
+import org.junit.runners.Parameterized.Parameters;
 
+//TODO: The IGJ has a completely different testing mechanism from the rest of the framework
+//TODO: We should probably standardize
 // Also see file FailedTests, that contains currently-failing tests.
 /**
  * JUnit tests for the IGJ annotation checker.
  */
-public class IGJTest extends CheckerTest {
+public class IGJTest {
```
Comment (JonathanBurke) [inline checker/tests/src/tests/IGJTest.java]: The IGJ Checker was the only test that didn't fit the usual mold.  It required a few of the old methods from the test framework which have been moved here.  It should be updated to match our testing conventions.

File: checker/build.xml [hunk 2]
```diff
@@ -510,6 +520,15 @@
     <!-- Used to guarantee that targets executed from other ant scripts do not cause unnecessary rebuilds -->
     <target name="all-tests-nojar" depends="build-tests"
             description="Run tests for all checkers, WITHOUT building anything">
+
```
Comment (wmdietl) [inline checker/build.xml]: Can we reduce the duplication, which existed before your changes, and somehow merge the two "junit" tasks?

File: checker/tests/src/tests/FenumTest.java [hunk 3]
```diff
@@ -16,7 +17,7 @@ public FenumTest(File testFile) {
     }
 
     @Parameters
-    public static Collection<Object[]> data() {
-        return testFiles("fenum", "all-systems");
+    public static Collection<Object[]> getTestFiles() {
```
Comment (wmdietl) [inline checker/tests/src/tests/FenumTest.java]: This method repeats a lot, in each test file, and only the directory names differ.
To reduce this duplication and possible source of mistakes, could we introduce a

String[] getTestDirectories()

method? Then one would only need to write

return {"fenum", "all-systems"};

(If array literals work in that context...)

File: checker/tests/src/tests/IGJTest.java (hunk 1 above)
Comment (wmdietl) [inline checker/tests/src/tests/IGJTest.java]: I agree that the IGJ tests should be adapted to the current style - in particular merging all the .out files into the test .java files.

File: framework/src/org/checkerframework/framework/test/DefaultCheckerTest.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+import static org.checkerframework.framework.test.TestConfigurationBuilder.buildDefaultConfiguration;
+
+/**
+ * To use this class you must do two things:
+ * 1) Create exactly 1 constructor in the subclass with exactly 1 argument which is a java.io.File.  This File
+ * will be the Java file that is compiled and whose output is verified.
+ * 2) Create a public static method that is annotation with org.junit.runners.Parameterized.Parameters.  This method
+ * should look like:
+ * {@code, @Parameters List<Object[]> getTestFiles(); }
+ * The return type should be a List of 1 element arrays, where each member of that array is a Java file to be tested.
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/test/DefaultCheckerTest.java]: Why is it a "List of 1 element arrays", but then "each member of that array"?

File: framework/src/org/checkerframework/framework/test/DefaultCheckerTest.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+ * Usually the body of this method should look like:
+ * {@code, return findNestedJavaTestFiles("myCheckerDir1","all-systems","someDir"); }
+ *
+ * findNestedJavaTestFiles will prepend "path to currentWorkingDir/tests/" to the argument.
+ *
+ * The TestSuite will then instantiate the subclass once for each file returned by getTestFiles and execute
+ * the run method.
+ */
+@RunWith(TestSuite.class)
+public abstract class DefaultCheckerTest {
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/test/DefaultCheckerTest.java]: Should this be CheckerFrameworkTest?
Why the "Default" if it doesn't implement a "CheckerTest" interface?

File: framework/src/org/checkerframework/framework/test/DefaultCheckerTest.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    protected final String checkerDir;
+
+    /** Extra options to pass to javac when running the checker. */
+    protected final List<String> checkerOptions;
+
+    /**
+     * Creates a new checker test.
+     *
+     * @param checker the class for the checker to use
+     * @param checkerDir the path to the directory of test inputs
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/test/DefaultCheckerTest.java]: Why is the "checkerDir" parameter needed? Don't we get the test inputs from the "getTestFiles" method?

File: framework/src/org/checkerframework/framework/test/ImmutableTestConfiguration.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Represents all of the information needed to execute the Javac compiler for a given set of test files.
+ */
+public class ImmutableTestConfiguration implements TestConfiguration {
+
+    /**
+     * Options that should be passed to the compiler
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/test/ImmutableTestConfiguration.java]: Why is it a Map? What are the keys and values?

File: framework/src/org/checkerframework/framework/test/SimpleOptionMap.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Created by jburke on 7/7/15.
+ */
+public class SimpleOptionMap {
+    private Map<String, String> options = new LinkedHashMap<>();
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/test/SimpleOptionMap.java]: Field can be final?

File: framework/src/org/checkerframework/framework/test/SimpleOptionMap.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+import org.checkerframework.framework.util.PluginUtil;
+
+import java.io.File;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Created by jburke on 7/7/15.
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/test/SimpleOptionMap.java]: Some documentation would be useful.

File: framework/src/org/checkerframework/framework/test/TestConfigurationBuilder.java [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+    public TestConfigurationBuilder() {
+        diagnosticFiles = new ArrayList<>();
+        testSourceFiles = new ArrayList<>();
+        processors = new LinkedHashSet<>();
+        options = new SimpleOptionMap();
+        shouldEmitDebugInfo = false;
+    }
+
+    public <T extends Collection<E>, E>  T addAll(T coll, Iterable<? extends E> objs) {
+        for(E obj : objs) {
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/test/TestConfigurationBuilder.java]: Spacing in this file should consistently use "for (".

File: framework/src/org/checkerframework/framework/test/TestConfigurationBuilder.java [hunk 11]
```diff
@@ -0,0 +1,363 @@
...
+
+    public TestConfigurationBuilder() {
+        diagnosticFiles = new ArrayList<>();
+        testSourceFiles = new ArrayList<>();
+        processors = new LinkedHashSet<>();
+        options = new SimpleOptionMap();
+        shouldEmitDebugInfo = false;
+    }
+
+    public <T extends Collection<E>, E>  T addAll(T coll, Iterable<? extends E> objs) {
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/test/TestConfigurationBuilder.java]: Should this method be static? private?
The method modifies the first parameter. None of the invocations of the method that I see in this diff use the returned value. It would be clearer to return void and document the modification of the parameter.

File: framework/src/org/checkerframework/framework/test/TestConfigurationBuilder.java [hunk 12]
```diff
@@ -0,0 +1,363 @@
...
+               classpath + File.pathSeparator +
+               globalclasspath;
+    }
+
+    public static String getJdkJarPathFromProperty() {
+        return System.getProperty("JDK_JAR");
+    }
+
+    /**
+     * This is creates a builder for the default configuration used by Checker Framework JUnit tests.
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/test/TestConfigurationBuilder.java]: Remove "is".

File: framework/src/org/checkerframework/framework/test/TestConfigurationBuilder.java [hunk 13]
```diff
@@ -0,0 +1,363 @@
...
+
+
+    private List<File> diagnosticFiles;
+    private List<File> testSourceFiles;
+    private Set<String> processors;
+    private SimpleOptionMap options;
+
+    private boolean shouldEmitDebugInfo;
+
+    public TestConfigurationBuilder() {
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/test/TestConfigurationBuilder.java]: Should this constructor be private to force usage of the static factory methods?
Should those factory methods be moved to the top of the file to make them easier to find?

File: framework/src/org/checkerframework/framework/test/TestConfigurationBuilder.java [hunk 14]
```diff
@@ -0,0 +1,363 @@
...
+
+    public <T extends Collection<E>, E>  T addAll(T coll, Iterable<? extends E> objs) {
+        for(E obj : objs) {
+            coll.add(obj);
+        }
+
+        return coll;
+    }
+
+    public TestConfigurationBuilder(TestConfiguration initialConfig) {
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/test/TestConfigurationBuilder.java]: This constructor doesn't seem to be used. Should it be removed? private?

File: checker/tests/src/tests/OIGJTest.java [hunk 15]
```diff
@@ -21,8 +22,8 @@ public OIGJTest(File testFile) {
     }
 
     @Parameters
-    public static Collection<Object[]> data() {
-        return filter(testFiles("oigj", "all-systems"));
+    public static Collection<Object[]> getTestFiles() {
+        return filter(TestUtilities.findNestedJavaTestFiles("oigj", "all-systems"));
     }
 
     // Duplicate from JavariTest.
```
Comment (wmdietl) [inline checker/tests/src/tests/OIGJTest.java]: Is it worth moving the filter functionality into TestUtilities?
No, actually we should remove that code here and skip-test the files that are filtered.

**Comment by wmdietl** [comment]
Thanks for tackling this beast! I reviewed about the first half of the changes. I will continue with the rest some other time, maybe after you updated it.

File: checker/build.xml (hunk 2 above)
Comment (JonathanBurke) [inline checker/build.xml]: I agree with this.  I just haven't had time since I am trying to focus on the Verigames experiment we're trying to run.  I'll see if this can be done quickly but if not, I will add an issue and move on.

File: checker/tests/src/tests/OIGJTest.java (hunk 15 above)
Comment (JonathanBurke) [inline checker/tests/src/tests/OIGJTest.java]: I agree.  Perhaps we can put an issue to fix both the IGJ/OIGJ tests together and somebody can handle that when they get a chance.  However, I worry that it's low priority so it might never get done.  Then again, at least this would be an easy task for a new hire, one in which they could make forward progress right away.

File: checker/tests/src/tests/FenumTest.java (hunk 3 above)
Comment (JonathanBurke) [inline checker/tests/src/tests/FenumTest.java]: That's a good suggestion.  I have done the following:

DefaultCheckerTest nows asks you to specify one of two methods:
```
@Parameters
public static void List<Object[]> getTestFiles() 
```
Which does the old behavior and allows people to find arbitrary test files (which is used in the InferenceTest suite to find all-systems in the checker-framework directory).
OR users can specify
```
@Parameters
public static void String[] getTestDirs() 
```
Which will return a list of directories that should be in the test directory.  The TestSuite will now enforce that a single method with one of the exact signatures above and an `@Parameters` has been added to your test file and otherwise it will explode with an informative exception/message.

Comment (JonathanBurke) [inline checker/tests/src/tests/FenumTest.java]: Note, I think in the long run we should separated the tests and their corresponding input test files into two directories:
test 
testdata

I think it's confusing to have both the testdata and the tests that run against that data in the same folder at the same level.

File: framework/src/org/checkerframework/framework/test/DefaultCheckerTest.java (hunk 5 above)
Comment (JonathanBurke) [inline framework/src/org/checkerframework/framework/test/DefaultCheckerTest.java]: Well default in the sense that this is the default class users will want to extend.  Whereas, if you had special behavior you could just not extend this class and create your own `@Test` methods.  For instance, we also have an InferenceCheckerTest.

File: framework/src/org/checkerframework/framework/test/DefaultCheckerTest.java (hunk 4 above)
Comment (JonathanBurke) [inline framework/src/org/checkerframework/framework/test/DefaultCheckerTest.java]: Well I guess it should be:
I can update this comment. 
Alternatively:  I kept this signature because it matches the original framework.  We could make this a List<String> and do the conversion to a List<Object []> in the TestSuite to make this less ambiguous.    It needs to be List<Object []> because that's what JUnit expects for parameters.

File: framework/src/org/checkerframework/framework/test/ImmutableTestConfiguration.java (hunk 7 above)
Comment (JonathanBurke) [inline framework/src/org/checkerframework/framework/test/ImmutableTestConfiguration.java]: Some options have arguments and some do not, e.g.,
-AprintAllQualifiers
vs.
-Aclasspath "dir1:dir2"
While you can argue that second case is more like 2 arguments, the point of the map is so I can do things like:

addToClasspath("dir1")
addToClasspath("dir2")

rather than
setClasspath("dir1")
setClasspath(getClasspath() + File.pathSeparator" + "dir2")

It is a pretty naive approach to adding options but works pretty well in most cases.  Since we are using it in a limited set of cases in the test suite I am not sure we need to make it any more sophisticated.

Comment (JonathanBurke) [inline framework/src/org/checkerframework/framework/test/ImmutableTestConfiguration.java]: Though I should probably change this to SimpleOptionMap

File: framework/src/org/checkerframework/framework/test/SimpleOptionMap.java (hunk 9 above)
Comment (JonathanBurke) [inline framework/src/org/checkerframework/framework/test/SimpleOptionMap.java]: Yeah, sorry.  I was just butting up against inference deadlines but I think I can add that now as that was pushed back a bit (but is still a top priority).

File: framework/src/org/checkerframework/framework/test/ImmutableTestConfiguration.java (hunk 7 above)
Comment (JonathanBurke) [inline framework/src/org/checkerframework/framework/test/ImmutableTestConfiguration.java]: I take that back.  The reason this is a map is because of the above reasoning but the ImmutableTestConfiguration should not contain the SimpleOptionMap (because it is mutable).

File: framework/src/org/checkerframework/framework/test/TestConfigurationBuilder.java (hunk 11 above)
Comment (JonathanBurke) [inline framework/src/org/checkerframework/framework/test/TestConfigurationBuilder.java]: This method should actually be removed.  There is one below that is private static that is used instead.

Comment (JonathanBurke) [inline framework/src/org/checkerframework/framework/test/TestConfigurationBuilder.java]: That method, catIterableAndList has had its name changed to catListAndIterable

File: framework/src/org/checkerframework/framework/test/TestConfigurationBuilder.java (hunk 13 above)
Comment (JonathanBurke) [inline framework/src/org/checkerframework/framework/test/TestConfigurationBuilder.java]: I did not plan on forcing people to use the static methods.  It can be used either way.  I just added those methods in order to make it easy to cover the most common case.  I can move them up top, however.

File: framework/src/org/checkerframework/framework/test/TestConfigurationBuilder.java (hunk 14 above)
Comment (JonathanBurke) [inline framework/src/org/checkerframework/framework/test/TestConfigurationBuilder.java]: This constructor is used in the Inference Framework.

File: framework/src/org/checkerframework/framework/test/TestConfigurationBuilder.java (hunk 13 above)
Comment (JonathanBurke) [inline framework/src/org/checkerframework/framework/test/TestConfigurationBuilder.java]: Thinking this over, I think moving those static methods to the top makes the

**Comment by JonathanBurke** [comment]
Hi Werner, I have incorporated most of your feedback (and my subsequent replies), added more documentation, and removed a couple of superfluous methods.  Please let me know if you have any other suggestions.

File: checker/build.xml (hunk 2 above)
Comment (wmdietl) [inline checker/build.xml]: How about adding a comment to the two duplicates to make sure anyone who edits one also edits the other?
And adding an issue that we should fix the duplication.

File: framework/src/org/checkerframework/framework/test/ImmutableTestConfiguration.java (hunk 7 above)
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/test/ImmutableTestConfiguration.java]: You should add documentation for what the key and the values are.
I didn't make the connection from arguments to parameters being the values that the parameters map to.

File: checker/build.xml [hunk 16]
```diff
@@ -521,13 +540,15 @@
       <!-- Copied from -run-tests target -->
       <mkdir dir="${build.reports}"/>
 
-      <junit fork="${run.tests.should.fork}"
+        <junit fork="${run.tests.should.fork}"
```
Comment (wmdietl) [inline checker/build.xml]: Unnecessary whitespace change.

File: framework/src/org/checkerframework/framework/test/DefaultCheckerTest.java [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+ * The method returns a List of Java files. There are methods in TestUtilities like findNestedJavaTestFiles to help
+ * you construct this List. The TestSuite will then instantiate the subclass once for each file returned by
+ * getTestFiles and execute the run method.
+ * An example of this method is:
+ *
+ * {@code @Parameters public static List<Object[]> getTestFiles() { return TestUtilities.findNestedJavaTestFiles("aggregate"); } }
+ *
+ */
+@RunWith(TestSuite.class)
+public abstract class DefaultCheckerTest {
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/test/DefaultCheckerTest.java]: I would still call it CheckerFrameworkTest, because it's not a default implementation for some interface.
But this is OK. (I saw a comment from you on this in email, but not here...)

File: framework/src/org/checkerframework/framework/test/DefaultCheckerTest.java (hunk 6 above)
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/test/DefaultCheckerTest.java]: Can we remove the parameter?

File: framework/src/org/checkerframework/framework/test/TestConfiguration.java [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+import javax.tools.JavaCompiler;
+import javax.tools.JavaFileManager;
+import javax.tools.JavaFileObject;
+import javax.tools.StandardJavaFileManager;
+import java.io.File;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Created by jburke on 6/24/15.
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/test/TestConfiguration.java]: A one sentence summary of the purpose of the interface would be more useful.

File: framework/src/org/checkerframework/framework/test/TestSuite.java [hunk 19]
```diff
@@ -0,0 +1,200 @@
...
+
+        int modifiers = method.getMethod().getModifiers();
+        if (!Modifier.isStatic(modifiers) || ! Modifier.isPublic(modifiers)) {
+            throw new RuntimeException("Parameter method (" + method.getName() +") must be public and static");
+        }
+
+        return method;
+    }
+
+    private static String requiredFormsMessage =
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/test/TestSuite.java]: Can be final?

**Comment by wmdietl** [comment]
Great job! Having getTestDirs makes tests so much simpler!
I think we should pull this and get it into the upcoming release.

File: framework/src/org/checkerframework/framework/test/ImmutableTestConfiguration.java (hunk 7 above)
Comment (JonathanBurke) [inline framework/src/org/checkerframework/framework/test/ImmutableTestConfiguration.java]: SimpleOptionMap is a utility class that contains a Map like above and a set of utility methods that manipulate the map.  SimpleOptionMap contains methods similar to addToClasspath above except they work like:  addToPath("-cp", "myClassesDir").  Then the ConfigurationBuilder's method looks as follows:
public void addToClasspath(String classesDir) {
   optionMap.addToPath("-cp", "myClassesDir");
}

Comment (JonathanBurke) [inline framework/src/org/checkerframework/framework/test/ImmutableTestConfiguration.java]: One final note, I have been meaning to add this documentation to the interface NOT the implementation, so most of these methods will just `@inherit` the Javadoc.

File: framework/src/org/checkerframework/framework/test/DefaultCheckerTest.java (hunk 6 above)
Comment (JonathanBurke) [inline framework/src/org/checkerframework/framework/test/DefaultCheckerTest.java]: So checkerDir is used right now to put on the sourcepath for the JUnit test.  We probably can but I think I will avoid doing this for now as we're getting close to release and I have many tasks.  Perhaps when open the issue for separating the Testdata I will also add information about removing this parameter.

File: checker/build.xml (hunk 16 above)
Comment (JonathanBurke) [inline checker/build.xml]: the spacing on this target is actually off compared to the others in the doc.  I will make it consistent.
